src/java/org/apache/cassandra/index/sai/disk/vector/VectorIndexSearcher.java

```diff
@@ -36,40 +38,46 @@
 import org.apache.cassandra.index.sai.disk.PostingList;
 import org.apache.cassandra.index.sai.disk.PrimaryKeyMap;
 import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
-import org.apache.cassandra.index.sai.disk.hnsw.CassandraOnDiskHnsw;
+import org.apache.cassandra.index.sai.disk.format.Version;
+import org.apache.cassandra.index.sai.disk.v1.IndexSearcher;
+import org.apache.cassandra.index.sai.disk.v1.PerIndexFiles;
+import org.apache.cassandra.index.sai.disk.v1.SegmentMetadata;
 import org.apache.cassandra.index.sai.disk.v1.postings.ReorderingPostingList;
+import org.apache.cassandra.index.sai.disk.v2.hnsw.CassandraOnDiskHnsw;
+import org.apache.cassandra.index.sai.disk.v3.CassandraDiskAnn;
 import org.apache.cassandra.index.sai.plan.Expression;
 import org.apache.cassandra.index.sai.utils.ArrayPostingList;
 import org.apache.cassandra.index.sai.utils.PrimaryKey;
 import org.apache.cassandra.index.sai.utils.RangeIterator;
 import org.apache.cassandra.index.sai.utils.RangeUtil;
 import org.apache.cassandra.index.sai.utils.SegmentOrdering;
-import org.apache.lucene.util.Bits;
-import org.apache.lucene.util.SparseFixedBitSet;
 
 /**
- * Executes ann search against the HNSW graph for an individual index segment.
+ * Executes ann search against the graph for an individual index segment.
  */
 public class VectorIndexSearcher extends IndexSearcher implements SegmentOrdering
 {
     private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
 
-    private final CassandraOnDiskHnsw graph;
+    private final JVectorLuceneOnDiskGraph graph;
     private final PrimaryKey.Factory keyFactory;
-    private final VectorType<float[]> type;
     private int maxBruteForceRows; // not final so test can inject its own setting
     private final ThreadLocal<SparseFixedBitSet> cachedBitSets;
 
-    VectorIndexSearcher(PrimaryKeyMap.Factory primaryKeyMapFactory,
-                        PerIndexFiles perIndexFiles,
-                        SegmentMetadata segmentMetadata,
-                        IndexDescriptor indexDescriptor,
-                        IndexContext indexContext) throws IOException
+    public VectorIndexSearcher(PrimaryKeyMap.Factory primaryKeyMapFactory,
+                               PerIndexFiles perIndexFiles,
+                               SegmentMetadata segmentMetadata,
+                               IndexDescriptor indexDescriptor,
+                               IndexContext indexContext) throws IOException
     {
         super(primaryKeyMapFactory, perIndexFiles, segmentMetadata, indexDescriptor, indexContext);
-        graph = new CassandraOnDiskHnsw(segmentMetadata.componentMetadatas, perIndexFiles, indexContext);
+        if (indexDescriptor.version == Version.BA)
```

Comment:
This should really be integrated into the `OnDiskFormat` such that `IndexSearcher` becomes an interface and the `OnDiskFormat` provides the implementation. We could move the `open` method to the `OnDiskFormat` as `newIndexSearcher` and implement that for each of the versions. With a `V2VectorIndexSearcher` and `V3VectorIndexSearcher`.

I have pushed a commit here https://github.com/datastax/cassandra/tree/diskann3-index-searcher that implements this.

Reply:
Big improvement, thanks!  Merged into diskann3.